lib_src/src/asrc_task/asrc_task.c

```diff
@@ -0,0 +1,402 @@
+// Copyright 2024 XMOS LIMITED.
+// This Software is subject to the terms of the XMOS Public Licence: Version 1.
+#include <xcore/chanend.h>
+#include <xcore/parallel.h>
+#include <xcore/assert.h>
+#include <xcore/hwtimer.h>
+#include <xcore/interrupt.h>
+#include <xcore/interrupt_wrappers.h>
+#include <xcore/triggerable.h>
+
+#include <stdio.h>
+#include <print.h>
+
+#include <stdbool.h>
+#include <string.h>
+
+#include "asrc_task.h"
+#include "asrc_timestamp_interpolation.h"
+
+#ifdef DEBUG_ASRC_TASK
+#define dprintf(...)   printf(__VA_ARGS__)
+#else
+#define dprintf(...)
+#endif
+
+
+static int frequency_to_fs_code(int frequency) {
+    if(frequency == 44100) {
+        return  FS_CODE_44;
+    } else if(frequency == 48000) {
+        return  FS_CODE_48;
+    } else if(frequency == 88200) {
+        return  FS_CODE_88;
+    } else if(frequency == 96000) {
+        return  FS_CODE_96;
+    } else if(frequency == 176400) {
+        return  FS_CODE_176;
+    } else if(frequency == 192000) {
+        return  FS_CODE_192;
+    } else {
+        xassert(0);
+    }
+    return -1;
+}
+
+
+// Structure used for thread scheduling of parallel ASRC
+typedef struct schedule_info_t{
+    int num_channels;
+    int channel_start_idx;
+} schedule_info_t;
+
+
+// Generates a schedule based on the number of channels of ASRC needed vs available threads
+int calculate_job_share(int asrc_channel_count,
+                        schedule_info_t *schedule){
+    int channels_per_first_job = (asrc_channel_count + MAX_ASRC_THREADS - 1) / MAX_ASRC_THREADS; // Rounded up channels per max jobs
+    int channels_per_last_job = 0;
+
+    int num_jobs = 0;
+    while(asrc_channel_count > 0){
+        channels_per_last_job = asrc_channel_count;
+        asrc_channel_count -= channels_per_first_job;
+        schedule[num_jobs].num_channels = channels_per_first_job;
+        schedule[num_jobs].channel_start_idx = channels_per_first_job * num_jobs;
+        num_jobs++;
+    };
+    schedule[num_jobs - 1].num_channels = channels_per_last_job;
+
+    return num_jobs;
+}
+
+
+// A single worker thread which operates on a group of channels in parallel with other worker threads
+DECLARE_JOB(do_asrc_group, (schedule_info_t*, uint64_t, asrc_in_out_t*, unsigned, int*, asrc_ctrl_t*));
+void do_asrc_group(schedule_info_t *schedule, uint64_t fs_ratio, asrc_in_out_t * asrc_io, unsigned input_write_idx, int* num_output_samples, asrc_ctrl_t asrc_ctrl[]){
+
+    // Make copies for readability
+    int num_worker_channels = schedule->num_channels;
+    int worker_channel_start_idx = schedule->channel_start_idx;
+
+    // Pack into the frame this instance of ASRC expects
+    int input_samples[ASRC_N_IN_SAMPLES * MAX_ASRC_CHANNELS_TOTAL];
+    for(int i = 0; i < ASRC_N_IN_SAMPLES * num_worker_channels; i++){
+        int rd_idx = i % num_worker_channels + (i / num_worker_channels) * asrc_io->asrc_channel_count + worker_channel_start_idx;
```

Comment:
This is a great optimisation (also for wr_idx). I have left the original line in commented out with "optimisation" added to the new one. The cycles are probably not so important but div is a shared co-proc so it could have side-effects on other threads so worth doing.
I have added a multi-channel tone test to ensure signals are in correct channels (not that I don't trust your opt obvs...just a good piece of extended coverage!).